app/scripts/controllers/detect-tokens.js

```diff
@@ -24,71 +23,84 @@ export default class DetectTokensController {
     preferences,
     network,
     keyringMemStore,
+    tokenList,
   } = {}) {
     this.preferences = preferences;
     this.interval = interval;
     this.network = network;
     this.keyringMemStore = keyringMemStore;
+    this.tokenList = tokenList;
+  }
+
+  async _getTokenBalances(tokens) {
+    const ethContract = this.web3.eth
+      .contract(SINGLE_CALL_BALANCES_ABI)
+      .at(SINGLE_CALL_BALANCES_ADDRESS);
+    return new Promise((resolve, reject) => {
+      ethContract.balances([this.selectedAddress], tokens, (error, result) => {
+        if (error) {
+          return reject(error);
+        }
+        return resolve(result);
+      });
+    });
   }
 
   /**
-   * For each token in @metamask/contract-metadata, find check selectedAddress balance.
+   * For each token in tokenlist ptovided by the TokenListController, find check selectedAddress balance.
    */
   async detectNewTokens() {
     if (!this.isActive) {
       return;
     }
-    if (this._network.store.getState().provider.chainId !== MAINNET_CHAIN_ID) {
+    if (Object.keys(this._tokenList.state.tokenList).length === 0) {
       return;
     }
 
     const tokensToDetect = [];
     this.web3.setProvider(this._network._provider);
-    for (const contractAddress in contracts) {
+    const { tokenList } = this._tokenList.state;
+    for (const tokenAddress in tokenList) {
       if (
-        contracts[contractAddress].erc20 &&
-        !this.tokenAddresses.includes(contractAddress.toLowerCase()) &&
-        !this.hiddenTokens.includes(contractAddress.toLowerCase())
+        !this.tokenAddresses.find((address) =>
+          isEqualCaseInsensitive(address, tokenAddress),
+        ) &&
+        !this.hiddenTokens.find((address) =>
+          isEqualCaseInsensitive(address, tokenAddress),
+        )
       ) {
-        tokensToDetect.push(contractAddress);
+        tokensToDetect.push(tokenAddress.toLowerCase());
       }
     }
-
-    let result;
-    try {
-      result = await this._getTokenBalances(tokensToDetect);
-    } catch (error) {
-      warn(
-        `MetaMask - DetectTokensController single call balance fetch failed`,
-        error,
-      );
-      return;
-    }
-
-    tokensToDetect.forEach((tokenAddress, index) => {
-      const balance = result[index];
-      if (balance && !balance.isZero()) {
-        this._preferences.addToken(
-          tokenAddress,
-          contracts[tokenAddress].symbol,
-          contracts[tokenAddress].decimals,
+    const sliceOfTokensToDetect = [];
+    sliceOfTokensToDetect[0] = tokensToDetect.slice(0, 1000);
+    sliceOfTokensToDetect[1] = tokensToDetect.slice(
+      1000,
+      tokensToDetect.length - 1,
+    );
+    for (const tokensSlice of sliceOfTokensToDetect) {
+      let result;
+      try {
+        result = await this._getTokenBalances(tokensSlice);
```

Comment:
This could potentially be a huge list, correct?  Is there a reason we can't do `Promise.all` here?